### ISSUE TITLE
BatchNorm numerical instability

### ISSUE THREAD

**Issue by classner** [BVLC/caffe#3963]
I just tracked down an issue with unexpected BatchNorm behavior in the current caffe version. The effect can be condensed to the following scenario: for an input batch with a channel with constant values, it produces (depending on the magnitude of the values) quite different results from zero. The expected normalization would be zeros in this channel for all values (mean normalized to zeros, no variance).

To reproduce the effect, I created a test with a 1x1x3x3 input blob where all entries are 100 (also works for smaller values). In this case, the computed mean is 100.000015 and the result of the normalization is ~ -0.00483, which is far away from 0. The normalization factor increases the numerical deviation. This just highlights how numerically badly conditioned the operation is and how important it is to get the mean right (the variance is then computed as E((X-E(X))^2) ). One approach to improve could be the stable online mean and variance computation by [Knuth](https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Online_algorithm) applied in regions with a reduction. 

Interestingly, chainer and mxnet do not seem to suffer from this effect. Chainer uses a reduction kernel for the mean computation [here](https://github.com/pfnet/chainer/blob/457550fc2501f3b8f73f93d69400f51d4096696f/cupy/core/core.pyx#L2299) and mxnet like [this](https://github.com/dmlc/mxnet/blob/master/src/operator/batch_norm-inl.h#L102). Probably the new cudnn versions handle this similarly.

This raises two issues:

1. This should be improved in general and,
2. it might lead to inconsistencies in whether the cudnn implementation or the caffe implementation is used.

For reference:

* [CuDNN v4/5 PR](https://github.com/BVLC/caffe/pull/3919)
* [CuDNN BatchNorm difference issue](https://github.com/NVIDIA/DIGITS/issues/629)
* [Original BatchNorm PR](https://github.com/BVLC/caffe/pull/3229)

**Comment by borisgin** [comment]
I rewrote BN layer ( see  https://github.com/drnikolaev/caffe/tree/caffe-0.15 
or https://github.com/borisgin/caffe/tree/caffe-0.15). 
Do you still observe this instability in the new BN implementation?

**Comment by wlike** [comment]
@borisgin the implementation at https://github.com/borisgin/caffe/tree/caffe-0.15 still has the problem described by @classner

**Comment by wlike** [comment]
@borisgin sorry, the above comment is wrong. The instability problem is **NOT** exist in your implementation, but there are some tiny mistakes

**Comment by wlike** [comment]
@classner I did a unit test of BatchNormLayer according to your statement. According to the debug info, the computed mean is right. The deviation happens in the second `caffe_cpu_gemm` following "subtract mean" comment. It's weired.
![ut](https://cloud.githubusercontent.com/assets/6084282/14701858/0805386e-07da-11e6-811a-09dd999acfa8.png)
![ut_code](https://cloud.githubusercontent.com/assets/6084282/14701859/080cc0c0-07da-11e6-8e1b-7845e52d5f05.png)
![ut_log](https://cloud.githubusercontent.com/assets/6084282/14701860/0811f202-07da-11e6-9ccf-16abf207eee9.png)

**Comment by classner** [comment]
Thank you, @wlike, for looking into this. I wanted to post the code, but was quite busy last week and could not catch up. I had set up a near identical test case.

For me, the mean already was off. It also depends on where the normalization factor is multiplied in in the mean calculation (first gemv or second). The result remains approximately the same, though...

**Comment by borisgin** [comment]
I pushed a new version of  BN following the review. New code:
-  fixed check in Reshape,
-  add "varaiance clipping" 
- changed initlization of global mean/ and variance 
Most sigbnificant piece is "Variance clipping" . I added it  to improve the stability of original BN algorithm since plain grad clipping does not help:
// clip variance
    if ((this->phase_ == TRAIN) && (iter_ <= BN_VARIANCE_CLIP_START))
      iter_++;
    if (iter_ > BN_VARIANCE_CLIP_START) {
      // clip from above
      // temp_C_[c] = average_var + gobal_var[c]
      Dtype y = caffe_cpu_asum(C, this->blobs_[3]->cpu_data());
      caffe_cpu_scale(C, Dtype(y/C), ones_C_.cpu_data(),
          temp_C_.mutable_cpu_data());
      caffe_cpu_axpby(C, Dtype(1.0), this->blobs_[3]->cpu_data(),
          Dtype(1.0), temp_C_.mutable_cpu_data());
      caffe_cpu_eltwise_min(C,
          Dtype(BN_VARIANCE_CLIP_CONST), temp_C_.cpu_data(),
          Dtype(1.0), variance_.mutable_cpu_data());
      // clip from below
      caffe_cpu_eltwise_max(C,
          Dtype((1.)/BN_VARIANCE_CLIP_CONST), this->blobs_[3]->cpu_data(),
          Dtype(1.0), variance_.mutable_cpu_data());
    }
